cmd/controller/app/controllers.go

```diff
@@ -64,28 +68,55 @@ func addControllers(mgr manager.Manager, client k8s.Client, informerFactory info
-			informerFactory.KubeSphereSharedInformerFactory().Devops().V1alpha3().Pipelines())
-
 		devopsCredentialController = devopscredential.NewController(client.Kubernetes(),
 			devopsClient,
 			informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
 			informerFactory.KubernetesSharedInformerFactory().Core().V1().Secrets())
 
-		jenkinsConfigController = jenkinsconfig.NewController(&jenkinsconfig.ControllerOptions{
-			LimitRangeClient:    client.Kubernetes().CoreV1(),
-			ResourceQuotaClient: client.Kubernetes().CoreV1(),
-			ConfigMapClient:     client.Kubernetes().CoreV1(),
-
-			ConfigMapInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().ConfigMaps(),
-			NamespaceInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
-			InformerFactory:   informerFactory,
-
-			ConfigOperator:  devopsClient,
-			ReloadCasCDelay: s.JenkinsOptions.ReloadCasCDelay,
-		}, s.JenkinsOptions)
+		// Choose controllers of CRDs (Pipeline and PipelineRun),
+		// by the field `PipelineBackend`in options.DevOpsControllerManagerOptions
+		if s.PipelineBackend == "jenkins" {
+			klog.Info("Jenkins was chosen to be the pipeline backend.")
+			devopsPipelineController = pipeline.NewController(client.Kubernetes(),
+				client.KubeSphere(), devopsClient,
+				informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
+				informerFactory.KubeSphereSharedInformerFactory().Devops().V1alpha3().Pipelines())
+
+			jenkinsConfigController = jenkinsconfig.NewController(&jenkinsconfig.ControllerOptions{
+				LimitRangeClient:    client.Kubernetes().CoreV1(),
+				ResourceQuotaClient: client.Kubernetes().CoreV1(),
+				ConfigMapClient:     client.Kubernetes().CoreV1(),
+
+				ConfigMapInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().ConfigMaps(),
+				NamespaceInformer: informerFactory.KubernetesSharedInformerFactory().Core().V1().Namespaces(),
+				InformerFactory:   informerFactory,
+
+				ConfigOperator:  devopsClient,
+				ReloadCasCDelay: s.JenkinsOptions.ReloadCasCDelay,
+			}, s.JenkinsOptions)
+		} else if s.PipelineBackend == "tekton" {
+			klog.Info("Tekton was chosen to be the pipeline backend.")
+			// add tekton pipeline controller
+			if err := (&tPipeline.PipelineReconciler{
+				Client: mgr.GetClient(),
+				Scheme: mgr.GetScheme(),
+			}).SetupWithManager(mgr); err != nil {
+				klog.Errorf("unable to create tekton-pipeline-controller, err: %v", err)
+				os.Exit(1)
+			}
+
+			// add tekton pipelinerun controller
+			if err := (&tPipelineRun.PipelineRunReconciler{
+				Client: mgr.GetClient(),
+				Scheme: mgr.GetScheme(),
+			}).SetupWithManager(mgr); err != nil {
+				klog.Errorf("unable to create tekton-pipelinerun-controller, err: %v", err)
+				os.Exit(1)
+			}
+		} else {
+			klog.Errorf("Pipeline backend does not found. Expected jenkins or tekton, but given %s", s.PipelineBackend)
+			os.Exit(1)
```

Comment:
I prefer to return an error instead of making the program exit directly. The caller of this function can handle it better.

same suggestion:
* [line 105](https://github.com/kubesphere/ks-devops/pull/179/files#diff-28af9f52994b1815410761cebf2f9631ca5f0e721d10d5c07f9cc7012b5a43e0R105)
* [line 114](https://github.com/kubesphere/ks-devops/pull/179/files#diff-28af9f52994b1815410761cebf2f9631ca5f0e721d10d5c07f9cc7012b5a43e0R114)

Reply:
Yes, I agree that the caller can handle the error.

Reply:
> I prefer to return an error instead of making the program exit directly. The caller of this function can handle it better.
> 
> same suggestion:
> 
> * [line 105](https://github.com/kubesphere/ks-devops/pull/179/files#diff-28af9f52994b1815410761cebf2f9631ca5f0e721d10d5c07f9cc7012b5a43e0R105)
> * [line 114](https://github.com/kubesphere/ks-devops/pull/179/files#diff-28af9f52994b1815410761cebf2f9631ca5f0e721d10d5c07f9cc7012b5a43e0R114)

Yes, I agree that the caller can handle the error. So I will return an error instead.